tools/dotnet-linker/Steps/ManagedRegistrarStep.cs

```diff
@@ -459,12 +461,57 @@ public void EmitCallToExportedMethod (MethodDefinition method, MethodDefinition
 					// We're throwing an exception, so there's no need for any more code.
 					skipEverythingAfter = il.Body.Instructions.Last ();
 				} else {
-					il.Emit (OpCodes.Ldarg_0);
+					// Whenever there's an NSObject constructor that we call from a registrar callback, we need to create
+					// a separate constructor that will first set the `handle` and `flags` values of the NSObject before
+					// calling the original constructor. Here's an example of the code we generate:
+					//
+					// // The original constructor:
+					// public .ctor (T0 p0, T1 p1, ...) { /* ... */ }
+					//
+					// // The generated constructor with pre-initialization: 
+					// public .ctor (NSObjectInitializationData initData, T0 p0, T1 p1, ...) {
+					//     this.handle = initData.Handle;
+					//     this.flags = initData.Flags;
+					//     this..ctor (p0, p1, ...);
+					// }
+					//
+					// - This code can't be expressed in C# and it can only be expressed directly in IL.
+					// - The reason we need to do this is because the base NSObject parameterless constructor 
+					//   would allocate a new Objective-C object if `handle` is a zero pointer.
+					// - The `NSObjectInitializationData` type is generated at build time and so it can't be used
+					//   anywhere in the app. The generated ctors will have a unique signature.
+
+					if (initializationDataType is null) {
+						initializationDataType = CreateNSObjectInitializationDataStruct ();
+						abr.PlatformAssembly.MainModule.Types.Add (initializationDataType.Type.Resolve ());
```

Comment:
I think it's possible to avoid the initialization data type by:

1. The flags value is constant, so it doesn't need to be passed around.
2. Add a IntPtr/NativeHandle parameter to pass the handle.
3. Add a dummy parameter with a non-public type (ObjCRuntime.RegistrarHelper comes to mind, since we already switch this to be public) to ensure the signature is unique, and then just pass `null` when invoking the constructor.